### DISCUSSION TITLE
Signed Windows Installer

### DISCUSSION THREAD

**Discussion by dlemstra** [ImageMagick/ImageMagick#6826] [Announcements]
Today our code signing certificate will expire. For many years [LeaderSSL](https://www.leaderssl.nl/) sponsored us with a code singing certificate but they are no longer able to do so. Since June of 2023 the CA/B Forum requires that OV code signing private keys be stored on a FIPS 140-2 Level 2 or Common Criteria Level EAL4+ certified device. This means we are no longer able to export our code signing certificate with its private key and use this in GitHub actions. We would now either need to have our own GitHub agent and hardware token or use a cloud solution (e.g. [digicert](https://docs.digicert.com/en/software-trust-manager/ci-cd-integrations/plugins/github-custom-action-for-keypair-signing.html)). Our preference would be to use a cloud solution that integrates with GitHub. Digicert seems to be our only option now but a certificate there would cost [$629](https://order.digicert.com/step1/code_signing) (tax excluded) for a single year. If your organization requires a signed installer then please consider sponsoring us with a code signing certificate. Please reach out to @dlemstra for questions or in case of a sponsorship.

**Comment by daiplusplus** [comment]
Does this only impact the .exe installer, or will all Win32 binaries (like the portable zip;s `magick.exe`) be unsigned?

**Reply by dlemstra** [reply]
This impacts all binaries that were signed with a code signing certificate.

**Comment by kevinmehall** [comment]
Have you looked into https://signpath.org/?

**Reply by dlemstra** [reply]
I was not aware of this company but that might be an option.

**Reply by Breakthrough** [reply]
+1, I have had a very positive experience using SignPath for my OSS projects.  I've been using them for over two years now, and they have consistently been fantastic to work with (always responsive when I have questions).

Note that you will need to use AppVeyor for building the signed binaries/installers due to the way they implement origin verification, but this is fairly easy to get going if you're already familiar with Github Actions.  I believe the installer also needs to be .MSI format.

**Reply by stefan-wenig** [reply]
Stefan from SignPath here. Thanks Brandon! <3 

Some restrictions apply to free certificates issued to "SignPath Foundation":
* We absolutely require transparent, verifiable builds for those, and currently this means AppVeyor. However, our GitHub actions connector will be available very soon. 
* While there are some advantages for standard formats like MSI, MSIX, AppX etc, we can also handle other installers. 

However, since ImageMagick has a registered LLC, we could get a certificate in your name with no restrictions. We can sponsor it for the first year, then you can either implement our verifiable process (on GHA) and get a free cert, or find another sponsor for the cert (<< $600) and keep using SignPath for free.

If that sounds good, just contact support@signpath.io.

**Reply by vico93** [reply]
I wonder if this could be used to sign drivers as well...

**Reply by stefan-wenig** [reply]
You would need a registered organization that can get both an EV certificate and be vetted by Microsoft for driver attestation signing.

SignPath cannot directly provide this. If your OSS project doesn't have an organization such as Image Magick's LLC or some foundation, you'll have to find a sponsor who trusts you with their name.

Sometimes project members or users have their own company or can convince their employer to step in.

The problem is that the certificate sponsor will be associates with any security breach in your drivers, or even be considered liable, so they often hesitate. SignPath can help by providing the platform that will allow your sponsor to control use of their certificate through effective OSS- friendly security policies.

**Comment by nicowilliams** [comment]
> Since June of 2023 the CA/B Forum requires that OV code signing private keys be stored on a FIPS 140-2 Level 2 or Common Criteria Level EAL4+ certified device. This means we are no longer able to export our code signing certificate with its private key and use this in GitHub actions.

Why can't your GHA reach out to your device via a proxy service you run and enable only when you're about to make releases, then disable when you're done making releases?

**Reply by dlemstra** [reply]
That would still require us to do a lot of extra work ourselves and that is why we are looking for a cloud based solution.

**Comment by yonathan06** [comment]
I use azure key vault with a certificate issued by global sign (EV code signing certificate), and then use AzureSignTool https://github.com/vcsjones/AzureSignTool to fetch the certificate and sign the file in github actions

**Reply by dlemstra** [reply]
Thanks for the information. That sounds like a cheaper option. And yesterday someone also pointed this tool out to me: https://github.com/dotnet/sign that uses AzureSignTools.

**Reply by riverar** [reply]
I moved from that to Azure Code Signing. Let me reach out to Microsoft and that team for you. Standby.

**Reply by dlemstra** [reply]
Thanks but I also just send out an email to AzureCodeSigningTAP myself to ask what would be possible.

**Reply by sudara** [reply]
This is what I do (GlobalSign + Azure Key Vault) — I wrote up an article on it https://melatonin.dev/blog/how-to-code-sign-windows-installers-with-an-ev-cert-on-github-actions/

**Reply by yonathan06** [reply]
here is the guide I was using: https://seankilleen.com/2020/05/how-to-use-azuresigntool-to-sign-files-with-azure-devops-using-a-certificate-in-azure-keyvault/
to install AzureSignTools in github actions:
```yml
      - if: matrix.platform == 'windows-latest'
        uses: actions/setup-dotnet@v3
        with:
          dotnet-version: '6.0.x'
      - if: matrix.platform == 'windows-latest'
        name: install AzureSignTools
        run: dotnet tool install --global azuresigntool
  ```

and then to sign:
```sh
azuresigntool sign -kvu ${process.env.AKV_URL} -kvi ${process.env.AKV_CLIENT_ID} -kvs ${process.env.AKV_CLIENT_SECRET} -kvt ${process.env.AKV_TENANT_ID} -kvc ${process.env.AKV_CERT_NAME} -tr ${tsServer} -ifl ${signFileListPath}
```
(process.env is JS access to env vars)

**Comment by jozefizso** [comment]
The Azure KeyVault supports this and it costs cents to sign builds several times a day.

**Reply by dlemstra** [reply]
The cost of a certificate certificate has still quadrupled in price but at least I now know options that are cheaper than digicert.

**Reply by jozefizso** [reply]
So mine certificate expired in Jan 2024. And now I have the same issue! The certs must be issued on a device and cannot be moved to Azure Key Vault. This is bummer.

**Comment by NicolasDorier** [comment]
hey look at this SignPath foundation seems to be a foundation offering free signing certificate. Unsure it works, never tried, but might be a path forward.

https://github.com/sigstore/fulcio/issues/250#issuecomment-1697130810

**Comment by Phillipus** [comment]
I don't want to hijack this discussion, but we are in the same position (https://github.com/archimatetool/archi/issues/990) so I would be interested in the outcome. FWIW I have written to the SignPath foundation to ask for a free certificate for our project. I'll update if I hear back from them.

**Comment by Jaxelr** [comment]
> I moved from that to Azure Code Signing. Let me reach out to Microsoft and that team for you. Standby.

hey @dlemstra , 

Engineer on Azure Code Signing here, feel free to reach out if you are not getting the right traction. I'm sure the team would be stoked to unblock you on this matter and we do support [Github Actions](https://github.com/marketplace/actions/azure-code-signing) as well.

Also adding our PM @ianjmcm

**Reply by dlemstra** [reply]
Thanks for your reply. I have already sent an e-mail to AzureCodeSigningTAP at microsoft.com this morning (local time) that will also probably go to ianjmcm but thanks for reaching out to me here.

**Reply by ValeraFinebits** [reply]
Hey @Jaxelr 
Could you please invite me to Azure Code Signing too?

**Reply by jozefizso** [reply]
hi @Jaxelr , our main OSS project https://github.com/NetOfficeFw is not stuck because our certificate expired on Jan 2024 and we cannot get a new one into Azure Key Vault. I already contacted the AzureCodeSigningTAP in Dec 2023 but was told I have to wait for public release.

Can we get access to Azure Code Signing for the NetOfficeFw project? We must release new bugfixes and the missing certificate does not allows us to publish nuget packages - they must be signed because the previous packages were signed.

**Comment by paulober** [comment]
Have you thought about self-signing the binaries and make the public certificate available for all to install once on their machines?

**Reply by dlemstra** [reply]
We have not thought about and we are now exploring the options that were offered here.

**Comment by dlemstra** [comment]
Thanks for all the options that were offered. We decided to use Azure Code Signing and this means we are now able to sign our binaries again. You can read here how easy it was to set this up here: https://github.com/dlemstra/github-stories/tree/main/2023/ImageMagick%20now%20uses%20Azure%20Code%20Signing.
